src/rpc/util.h

```diff
@@ -214,8 +209,11 @@ struct RPCArg {
 
     bool IsOptional() const;
 
-    /** Check whether the request JSON type matches. */
-    void MatchesType(const UniValue& request) const;
+    /**
+     * Check whether the request JSON type matches.
+     * Returns true if type matches, or object describing error(s) if not.
+     */
+    UniValue MatchesType(const UniValue& request) const;
```

Comment:
Feels like a `util::Result` return type would be more appropriate here?

Reply:
Not sure. This would be `util::Result<void, UniValue>`, which is not yet available. I could also use `std::optional<UniValue>`, but not sure if it is worth it.

Reply:
Yeah it would benefit from #25665 but could do it with today's implementation already as well. I didn't realize we need a `UniValue` object in `HandleRequest`, though, so that makes this a bit more verbose than I hoped. Just found it a bit confusing to have a `UniValue` act as a variant, but that's probably personal preference.

No strong view either way, just thinking out loud. Feel free to close.

<details>
<summary>git diff</summary>

```diff
diff --git a/src/rpc/util.cpp b/src/rpc/util.cpp
index 5f44b82aa..30528bd09 100644
--- a/src/rpc/util.cpp
+++ b/src/rpc/util.cpp
@@ -11,6 +11,7 @@
 #include <script/signingprovider.h>
 #include <tinyformat.h>
 #include <util/check.h>
+#include <util/result.h>
 #include <util/strencodings.h>
 #include <util/string.h>
 #include <util/system.h>
@@ -559,9 +560,9 @@ UniValue RPCHelpMan::HandleRequest(const JSONRPCRequest& request) const
     UniValue arg_mismatch{UniValue::VOBJ};
     for (size_t i{0}; i < m_args.size(); ++i) {
         const auto& arg{m_args.at(i)};
-        UniValue match{arg.MatchesType(request.params[i])};
-        if (!match.isTrue()) {
-            arg_mismatch.pushKV(strprintf("Position %s (%s)", i, arg.m_names), std::move(match));
+        auto match{arg.MatchesType(request.params[i])};
+        if (!match) {
+            arg_mismatch.pushKV(strprintf("Position %s (%s)", i, arg.m_names), UniValue{UniValue::VType::VSTR, util::ErrorString(match).original});
         }
     }
     if (!arg_mismatch.empty()) {
@@ -717,7 +718,7 @@ static std::optional<UniValue::VType> ExpectedType(RPCArg::Type type)
     NONFATAL_UNREACHABLE();
 }
 
-UniValue RPCArg::MatchesType(const UniValue& request) const
+util::Result<bool> RPCArg::MatchesType(const UniValue& request) const
 {
     if (m_opts.skip_type_check) return true;
     if (IsOptional() && request.isNull()) return true;
@@ -725,7 +726,7 @@ UniValue RPCArg::MatchesType(const UniValue& request) const
     if (!exp_type) return true; // nothing to check
 
     if (*exp_type != request.getType()) {
-        return strprintf("JSON value of type %s is not of expected type %s", uvTypeName(request.getType()), uvTypeName(*exp_type));
+        return util::Error{Untranslated(strprintf("JSON value of type %s is not of expected type %s", uvTypeName(request.getType()), uvTypeName(*exp_type)))};
     }
     return true;
 }
diff --git a/src/rpc/util.h b/src/rpc/util.h
index 0d60e29fa..c374ee528 100644
--- a/src/rpc/util.h
+++ b/src/rpc/util.h
@@ -16,6 +16,7 @@
 #include <script/standard.h>
 #include <univalue.h>
 #include <util/check.h>
+#include <util/result.h>
 
 #include <string>
 #include <variant>
@@ -213,7 +214,7 @@ struct RPCArg {
      * Check whether the request JSON type matches.
      * Returns true if type matches, or object describing error(s) if not.
      */
-    UniValue MatchesType(const UniValue& request) const;
+    util::Result<bool> MatchesType(const UniValue& request) const;
 
     /** Return the first of all aliases */
     std::string GetFirstName() const;
```
</details>

Reply:
Although I think with https://github.com/bitcoin/bitcoin/pull/25665 merged I'd definitely prefer having the `util::Result<void, UniValue>` return type, just for readability.

Reply:
Your diff doesn't work, because the return type is `util::Result<bool>`, but the function needs to return an `UniValue` object somehow. (To make the calling code less verbose and because it is needed for follow up 2/2)

Reply:
I think my diff does work? I'm not sure what the follow up and its requirements are, but the diff compiles, and seems to work fine e.g. for this call. But I agree that making the calling code less verbose is not ideal, so probably best to drop this approach.

<details>
<summary>./src/bitcoin-cli -signet getmempoolancestors tx 5</summary>

```
error code: -3
error message:
Wrong type passed:
{
    "Position 1 (verbose)": "JSON value of type number is not of expected type bool"
}
```
</summary>


I don't mean to bikeshed, but here's another approach without `util::Result` that may be more appropriate by returning `std::pair<bool, UniValue>`. Still slightly more verbose than currently, but could be worth it if compatible with 2/2? I'll give up after this :-D

<details>
<summary>git diff</summary>

```diff
diff --git a/src/rpc/util.cpp b/src/rpc/util.cpp
index 5f44b82aa..e0c6aec67 100644
--- a/src/rpc/util.cpp
+++ b/src/rpc/util.cpp
@@ -559,9 +559,9 @@ UniValue RPCHelpMan::HandleRequest(const JSONRPCRequest& request) const
     UniValue arg_mismatch{UniValue::VOBJ};
     for (size_t i{0}; i < m_args.size(); ++i) {
         const auto& arg{m_args.at(i)};
-        UniValue match{arg.MatchesType(request.params[i])};
-        if (!match.isTrue()) {
-            arg_mismatch.pushKV(strprintf("Position %s (%s)", i, arg.m_names), std::move(match));
+        auto [match, error] = arg.MatchesType(request.params[i]);
+        if (!match) {
+            arg_mismatch.pushKV(strprintf("Position %s (%s)", i, arg.m_names), std::move(error.value()));
         }
     }
     if (!arg_mismatch.empty()) {
@@ -717,17 +717,17 @@ static std::optional<UniValue::VType> ExpectedType(RPCArg::Type type)
     NONFATAL_UNREACHABLE();
 }
 
-UniValue RPCArg::MatchesType(const UniValue& request) const
+std::pair<bool, std::optional<UniValue>> RPCArg::MatchesType(const UniValue& request) const
 {
-    if (m_opts.skip_type_check) return true;
-    if (IsOptional() && request.isNull()) return true;
+    if (m_opts.skip_type_check) return {true, std::nullopt};
+    if (IsOptional() && request.isNull()) return {true, std::nullopt};
     const auto exp_type{ExpectedType(m_type)};
-    if (!exp_type) return true; // nothing to check
+    if (!exp_type) return {true, std::nullopt}; // nothing to check
 
     if (*exp_type != request.getType()) {
-        return strprintf("JSON value of type %s is not of expected type %s", uvTypeName(request.getType()), uvTypeName(*exp_type));
+        return {false, strprintf("JSON value of type %s is not of expected type %s", uvTypeName(request.getType()), uvTypeName(*exp_type))};
     }
-    return true;
+    return {true, std::nullopt};
 }
 
 std::string RPCArg::GetFirstName() const
diff --git a/src/rpc/util.h b/src/rpc/util.h
index 0d60e29fa..2d5f882fe 100644
--- a/src/rpc/util.h
+++ b/src/rpc/util.h
@@ -213,7 +213,7 @@ struct RPCArg {
      * Check whether the request JSON type matches.
      * Returns true if type matches, or object describing error(s) if not.
      */
-    UniValue MatchesType(const UniValue& request) const;
+    std::pair<bool, std::optional<UniValue>> MatchesType(const UniValue& request) const;
 
     /** Return the first of all aliases */
     std::string GetFirstName() const;
```
</details>

Reply:
The return value in 2/2 needs to be of a type that can hold the nested types of a JSON argument, so UniValue seems a good fit.

> std::pair<bool, UniValue>

yes, that works, but seems worse than just `std::optional<UniValue>`.

Reply:
> but seems worse than just std::optional<UniValue>

From a code brevity point, yes, but from a readability point of view I don't think so, because then boolean evaluation of `MatchesType` is the inverse of what the name implies.

Reply:
Ok, but then I'd prefer `UniValue` over `std::pair<bool, UniValue>`